optaplanner-examples/src/main/java/org/optaplanner/examples/nurserostering/optional/score/NurseRosteringConstraintProvider.java

```diff
@@ -0,0 +1,480 @@
+                                        breakInfo) -> breakInfo.getBreakLength() - 1 > employeeContractPair.getRight()
+                                                .getMaximumValue())
+                                .penalize("org.optaplanner.examples.nurserostering.solver", "maximumConsecutiveFreeDays",
+                                        HardSoftScore.ONE_SOFT,
+                                        (employeeContractPair, breakInfo) -> employeeContractPair.getRight()
+                                                .getViolationAmount(breakInfo.getBreakLength() - 1));
+    }
+
+    Constraint maximumConsecutiveFreeDaysNoAssignments(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(MinMaxContractLine.class)
+                .filter(minMaxContractLine -> minMaxContractLine
+                        .getContractLineType() == ContractLineType.CONSECUTIVE_FREE_DAYS &&
+                        minMaxContractLine.isMaximumEnabled())
+                .join(Employee.class, Joiners.equal(MinMaxContractLine::getContract, Employee::getContract))
+                .ifNotExists(ShiftAssignment.class,
+                        Joiners.equal((contract, employee) -> employee, ShiftAssignment::getEmployee))
+                .join(NurseRosterParametrization.class, Joiners.lessThan((contract, employee) -> contract.getMaximumValue(),
+                        nrp -> nrp.getLastShiftDateDayIndex() - nrp.getFirstShiftDateDayIndex() + 1))
+                .penalize("org.optaplanner.examples.nurserostering.solver", "maximumConsecutiveFreeDays (no shifts)",
+                        HardSoftScore.ONE_SOFT,
+                        (contract, employee, nrp) -> contract
+                                .getViolationAmount(nrp.getLastShiftDateDayIndex() - nrp.getFirstShiftDateDayIndex() + 1));
+    }
+
+    Constraint consecutiveFreeDaysFirstBreak(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(MinMaxContractLine.class)
+                .filter(minMaxContractLine -> minMaxContractLine
+                        .getContractLineType() == ContractLineType.CONSECUTIVE_FREE_DAYS &&
+                        minMaxContractLine.isEnabled())
+                .join(Employee.class, Joiners.equal(MinMaxContractLine::getContract, Employee::getContract))
+                .join(ShiftAssignment.class,
+                        Joiners.equal((contract, employee) -> employee, ShiftAssignment::getEmployee))
+                .groupBy((contract, employee, shiftAssignment) -> new BiTuple<>(contract, employee),
+                        ConstraintCollectors.<MinMaxContractLine, Employee, ShiftAssignment, ShiftAssignment> min(
+                                (contract, employee, shiftAssignment) -> shiftAssignment))
+                .join(NurseRosterParametrization.class)
+                .filter((contractEmployeePair, sa,
+                        nrp) -> nrp.getFirstShiftDate() != sa.getShiftDate() && contractEmployeePair.a
```

Comment:
If you need one reason to not use ImmutablePair here, it is this.
`contractEmployeePair.a` is very cryptic. Especially since it can just be `contract`.